_posts/2017-09-01-tips-for-upgrading-rails-3-2-to-4.markdown

```diff
@@ -0,0 +1,192 @@
+---
+layout: post
+title: "The definitive guide to upgrading from Rails 3.2 to 4.0"
+date: 2017-09-01 16:48:00
+categories: ["rails", "upgrades"]
+author: "mauro-oto"
+---
+
+_This article is part of our Upgrade Rails series. To see more of them, [click here](https://www.ombulabs.com/blog/tags/upgrades)_.
+
+A [previous post](https://www.ombulabs.com/blog/rails/tips-for-upgrading-rails-3-2-to-4.html)
+covered some general tips to take into account for this migration. This article
+will try to go a bit more in depth. We will first go from 3.2 to 4.0, and then
+to 4.2. Depending on the complexity of your app, a Rails upgrade can take
+anywhere from one week for a single developer, to a few months for two developers.
+
+1. Ruby version
+2. Gems
+3. Config files (config/)
+4. Application code
+  a. Models (app/models/)
+  b. Controllers (app/controllers/)
+5. Tests
+6. Miscellaneous
+7. Next steps
+
+## 1. Ruby version
+
+Rails 3.2.x is the last version to support Ruby 1.8.7. If you're using Ruby 1.8.7,
+you'll need to upgrade to Ruby 1.9.3 or newer. The Ruby upgrade is out of the scope
+of this guide, but check out [this guide](http://www.darkridge.com/~jpr5/2012/10/03/ruby-1.8.7-1.9.3-migration), which
+is very complete and serves as a sort of checklist, and we've used it in the past.
+
+## 2. Gems
+
+You can add the aptly named [rails4_upgrade gem](https://github.com/alindeman/rails4_upgrade)
+to your Rails 3 project's Gemfile and find gems which you'll need to update:
+
+```
+➜  myproject git:(develop) ✗ bundle exec rake rails4:check
+
+** GEM COMPATIBILITY CHECK **
++------------------------------------------+----------------------------+
+| Dependency Path                          | Rails Requirement          |
++------------------------------------------+----------------------------+
+| devise 2.1.4                             | railties ~> 3.1            |
+| devise-encryptable 0.2.0 -> devise 2.1.4 | railties ~> 3.1            |
+| friendly_id 4.0.10.1                     | activerecord < 4.0, >= 3.0 |
+| strong_parameters 0.2.3                  | actionpack ~> 3.0          |
+| strong_parameters 0.2.3                  | activemodel ~> 3.0         |
+| strong_parameters 0.2.3                  | activesupport ~> 3.0       |
+| strong_parameters 0.2.3                  | railties ~> 3.0            |
++------------------------------------------+----------------------------+
+```
+
+Instead of going through your currently bundled gems or `Gemfile.lock` manually,
+you get a report of the gems you need to upgrade.
+
+## 3. Config
+
+Rails includes the `rails:update` [task](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#the-update-task),
+which you can use as a guideline as explained thoroughly in
+[this post](http://thomasleecopeland.com/2015/08/06/running-rails-update.html)
+to get rid of unnecessary code or monkey-patches in your config files and
+initializers, specially if your Rails 3 app was running on Rails 2.
```

Comment:
@mauro-oto This sentence is quite long and hard to follow. Try breaking it down into smaller sentences.

```
Rails includes the `rails:update` [task](http://edgeguides.rubyonrails.org/upgrading_ruby_on_rails.html#the-update-task). You can use this task as a guideline as explained thoroughly in [this post](http://thomasleecopeland.com/2015/08/06/running-rails-update.html). It will help you get rid of unnecessary code or monkey-patches in your config files and initializers, especially if your Rails 3 app was running on Rails 2.
```